spi/imx6ull-qspi/qspi.c

```diff
@@ -0,0 +1,326 @@
+#define QSPI_RBSR_TO_READ (((*QSPI_RBSR_ADRR) >> 8) & 0x3F)
+
+#define QSPI_SFAR_ADDR (qspi_common.base + 0x100 / sizeof(uint32_t))
+
+#define QSPI_RBCT_ADDR  (qspi_common.base + 0x110 / sizeof(uint32_t))
+#define QSPI_RBCT_RXBRD (0x01 << 8)
+
+#define QSPI_FR_ADDR (qspi_common.base + 0x160 / sizeof(uint32_t))
+#define QSPI_FR_RBDF (0x01 << 16)
+
+#define QSPI_TBDR_ADDR (qspi_common.base + 0x154 / sizeof(uint32_t))
+
+#define QSPI_SFA1AD (qspi_common.base + 0x180 / sizeof(uint32_t))
+#define QSPI_SFA2AD (qspi_common.base + 0x184 / sizeof(uint32_t))
+#define QSPI_SFB1AD (qspi_common.base + 0x188 / sizeof(uint32_t))
+#define QSPI_SFB2AD (qspi_common.base + 0x18C / sizeof(uint32_t))
+
+#define WATERMARK 16
+
+#define NUM_LUT_SEQ 16
+
+typedef struct {
+	int mux;
+	char val;
+	char sion;
+} qspi_pctl_mux_t;
+
+typedef struct {
+	int pad;
+	char speed;
+	char dse;
+	char sre;
+} qspi_pctl_pad_t;
+
+qspi_pctl_mux_t qspi_pctl_mux[2][8] = {
+	{ { pctl_mux_nand_d7, 2, 0 }, { pctl_mux_nand_ale, 2, 0 }, { pctl_mux_nand_wp, 2, 0 }, { pctl_mux_nand_rdy, 2, 0 }, { pctl_mux_nand_ce0, 2, 0 }, { pctl_mux_nand_ce1, 2, 0 }, { pctl_mux_nand_cle, 2, 0 }, { pctl_mux_nand_dqs, 2, 0 } },
+	{ { pctl_mux_nand_re, 2, 0 }, { pctl_mux_nand_we, 2, 0 }, { pctl_mux_nand_d0, 2, 0 }, { pctl_mux_nand_d1, 2, 0 }, { pctl_mux_nand_d2, 2, 0 }, { pctl_mux_nand_d3, 2, 0 }, { pctl_mux_nand_d4, 2, 0 }, { pctl_mux_nand_d5, 2, 0 } },
+};
+
+qspi_pctl_pad_t qspi_pctl_pad[2][8] = {
+	{ { pctl_pad_nand_d7, 1, 4, 0 }, { pctl_pad_nand_ale, 1, 4, 0 }, { pctl_pad_nand_wp, 1, 4, 0 }, { pctl_pad_nand_rdy, 1, 4, 0 }, { pctl_pad_nand_ce0, 1, 4, 0 }, { pctl_pad_nand_ce1, 1, 4, 0 }, { pctl_pad_nand_cle, 1, 4, 0 }, { pctl_pad_nand_dqs, 1, 4, 0 } },
+	{ { pctl_pad_nand_re, 1, 4, 0 }, { pctl_pad_nand_we, 1, 4, 0 }, { pctl_pad_nand_d0, 1, 4, 0 }, { pctl_pad_nand_d1, 1, 4, 0 }, { pctl_pad_nand_d2, 1, 4, 0 }, { pctl_pad_nand_d3, 1, 4, 0 }, { pctl_pad_nand_d4, 1, 4, 0 }, { pctl_pad_nand_d5, 1, 4, 0 } },
+};
+
+
+struct {
+	volatile uint32_t *base;
+	uint32_t flash_base_addr[2];
+	bool init;
+} qspi_common = { .base = NULL, .init = false, .flash_base_addr = { QSPI_MMAP_BASE, QSPI_MMAP_BASE + QSPI_FLASH_A1_SIZE + QSPI_FLASH_A2_SIZE } };
+
+// TODO ustawić ile ohmów w rejestrze na flashu.
+// TODO find clock frequency - ~50Mhz
+
+static void set_mux(qspi_dev_t dev_no)
+{
+	platformctl_t ctl;
+	int i;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_iomux;
+
+	for (i = 0; i < 8; i++) {
+		ctl.iomux.mux = qspi_pctl_mux[dev_no][i].mux;
+		ctl.iomux.mode = qspi_pctl_mux[dev_no][i].val;
+		ctl.iomux.sion = qspi_pctl_mux[dev_no][i].sion;
+		platformctl(&ctl);
+	}
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_iopad;
+
+	for (i = 0; i < 8; i++) {
+		ctl.iopad.pad = qspi_pctl_pad[dev_no][i].pad;
+		ctl.iopad.speed = qspi_pctl_pad[dev_no][i].speed;
+		ctl.iopad.dse = qspi_pctl_pad[dev_no][i].dse;
+		ctl.iopad.sre = qspi_pctl_pad[dev_no][i].sre;
+		platformctl(&ctl);
+	}
+}
+
+static void set_clk()
```

Comment:
Make sure to check warnings 
```suggestion
static void set_clk(void)
```